launcher/modplatform/atlauncher/ATLPackInstallTask.cpp

```diff
@@ -1075,36 +1076,13 @@ void PackInstallTask::install()
 
 static Meta::Version::Ptr getComponentVersion(const QString& uid, const QString& version)
 {
-    auto vlist = APPLICATION->metadataIndex()->get(uid);
-    if (!vlist)
-        return {};
-
-    if (!vlist->isLoaded()) {
-        QEventLoop loadVersionLoop;
-        auto task = vlist->getLoadTask();
-        QObject::connect(task.get(), &Task::finished, &loadVersionLoop, &QEventLoop::quit);
-        if (!task->isRunning())
-            task->start();
-
-        loadVersionLoop.exec();
-    }
-
-    auto ver = vlist->getVersion(version);
-    if (!ver)
-        return {};
-
-    if (!ver->isLoaded()) {
-        QEventLoop loadVersionLoop;
-        ver->load(Net::Mode::Online);
-        auto task = ver->getCurrentTask();
-        QObject::connect(task.get(), &Task::finished, &loadVersionLoop, &QEventLoop::quit);
-        if (!task->isRunning())
-            task->start();
-
-        loadVersionLoop.exec();
-    }
+    QEventLoop ev;
+    auto task = APPLICATION->metadataIndex()->loadVersion(uid, version);
+    QObject::connect(task.get(), &Task::finished, &ev, &QEventLoop::quit);
+    task->start();
+    ev.exec();
```

Comment:
I've seen this particular new sequence more than three times, > 3 is the point where I usually look to factor it out to a function.

also the creation of a sub event loop this many times makes my skin itch. should we perhaps look to create a `loadVersionAsync` method that takes a callback?

Reply:
> I've seen this particular new sequence more than three times, > 3 is the point where I usually look to factor it out to a function.

Yeah it it is present in a few places but that is because previously the versions were partially loaded in sync( the first load where we read from files), now all that logic is handled async inside the task(when it succeeds the version is fully loaded).
I will consider moving all this inside one function but sadly in the places where it's used, I can't get rid of QEventLoop as they expect an immediate answer(and that needs to be refactored a lot to make the task load async).




> also the creation of a sub event loop this many times makes my skin itch. should we perhaps look to create a `loadVersionAsync` method that takes a callback?

I prefer the signal/task approach I return a task(object) and connect to the signals. That way I can chain them if needed in another ConcurrentTask.